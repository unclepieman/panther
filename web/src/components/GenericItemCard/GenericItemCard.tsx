/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { Box, Card, Flex, Icon, IconButton, Img, Text, TextProps, CardProps } from 'pouncejs';
import { slugify } from 'Helpers/utils';
import { Link as RRLink } from 'react-router-dom';

interface GenericItemCardLogoProps {
  src: string;
}

interface GenericItemCardValueProps {
  id?: string;
  label?: string;
  value: string | number | React.ReactElement;
}

interface GenericItemCardLinkProps {
  to: string;
}

interface GenericItemCardHeadingValueProps {
  value: string | number | React.ReactElement;
  label: string;
  labelFirst?: boolean;
  withDivider?: boolean;
}

interface GenericItemCardComposition {
  Logo: React.FC<GenericItemCardLogoProps>;
  Header: React.FC;
  Heading: React.FC<TextProps>;
  Body: React.FC;
  Link: React.FC<GenericItemCardLinkProps>;
  OptionsButton: React.ForwardRefExoticComponent<React.RefAttributes<HTMLButtonElement>>;
  Value: React.FC<GenericItemCardValueProps>;
  ValuesGroup: React.FC;
  HeadingValue: React.FC<GenericItemCardHeadingValueProps>;
  LineBreak: React.FC;
}

const GenericItemCard: React.FC<CardProps> & GenericItemCardComposition = ({
  children,
  isHighlighted = false,
}) => {
  return (
    <Card as="section" variant="dark" overflow="hidden">
      <Box backgroundColor={isHighlighted ? 'navyblue-600' : 'transparent'} p={4}>
        <Flex position="relative" height="100%">
          {children}
        </Flex>
      </Box>
    </Card>
  );
};

const GenericItemCardHeader: React.FC = ({ children }) => {
  return (
    <Flex as="header" align="flex-start" mb={2}>
      {children}
    </Flex>
  );
};

const GenericItemCardHeading: React.FC<TextProps> = ({ children, ...rest }) => {
  return (
    <Text as="h4" fontWeight="medium" mr="auto" maxWidth="60%" wordBreak="break-word" {...rest}>
      {children}
    </Text>
  );
};

const GenericItemCardHeadingValue: React.FC<GenericItemCardHeadingValueProps> = ({
  value,
  label,
  labelFirst,
  withDivider,
  ...rest
}) => {
  return (
    <>
      <Box fontSize="small" maxWidth="20%" truncated {...rest}>
        {labelFirst && (
          <Text mr={1} as="span" color="navyblue-100">
            {label}
          </Text>
        )}
        {value}
        {!labelFirst && (
          <Text ml={1} as="span" color="navyblue-100">
            {label}
          </Text>
        )}
      </Box>
      {withDivider && (
        <Text px={2} fontSize="small" as="span" color="navyblue-100">
          &bull;
        </Text>
      )}
    </>
  );
};

const GenericItemCardBody: React.FC = ({ children }) => {
  return (
    <Flex direction="column" justify="space-between" width={1}>
      {children}
    </Flex>
  );
};

const GenericItemCardValuesGroup: React.FC = ({ children }) => {
  return (
    <Flex wrap="wrap" spacing={8}>
      {children}
    </Flex>
  );
};

const GenericItemCardLogo: React.FC<GenericItemCardLogoProps> = ({ src }) => {
  return <Img nativeWidth={20} nativeHeight={20} mr={5} alt="Logo" src={src} />;
};

const GenericItemCardOptionsButton = React.forwardRef<HTMLButtonElement>(function OptionsButton(
  props,
  ref
) {
  return (
    <Box ml={2} mt={-1}>
      <IconButton
        variant="ghost"
        variantBorderStyle="circle"
        variantColor="navyblue"
        icon="more"
        size="small"
        aria-label="Toggle Options"
        {...props}
        ref={ref}
      />
    </Box>
  );
});

const GenericItemCardValue: React.FC<GenericItemCardValueProps> = ({ label, value, id }) => {
  const cardId = id || slugify(`${label}${value}`);

  return (
    <Flex direction="column" as="dl" mt={2}>
      {label && (
        <Box
          as="dt"
          aria-labelledby={cardId}
          color="gray-300"
          fontSize="2x-small"
          fontWeight="medium"
        >
          {label}
        </Box>
      )}
      <Box
        as="dd"
        display="inline-flex"
        flexGrow={1}
        aria-labelledby={cardId}
        fontSize="medium"
        fontWeight="medium"
        opacity={value ? 1 : 0.3}
        alignItems="center"
        minHeight={24}
      >
        {value || 'Not Set'}
      </Box>
    </Flex>
  );
};

const GenericItemCardLink: React.FC<GenericItemCardLinkProps> = ({ to, ...rest }) => {
  return (
    <RRLink to={to} {...rest}>
      <Flex
        justify="center"
        align="center"
        width={24}
        height={24}
        backgroundColor="navyblue-200"
        borderColor="navyblue-200"
        _hover={{ backgroundColor: 'blue-300', borderColor: 'blue-300' }}
        borderRadius="circle"
      >
        <Icon type="arrow-forward" size="x-small" />
      </Flex>
    </RRLink>
  );
};

const GenericItemCardLineBreak: React.FC = () => <Box flexBasis="100%" height={0} />;

GenericItemCard.Body = GenericItemCardBody;
GenericItemCard.Header = GenericItemCardHeader;
GenericItemCard.Link = React.memo(GenericItemCardLink);
GenericItemCard.Heading = GenericItemCardHeading;
GenericItemCard.Logo = GenericItemCardLogo;
GenericItemCard.OptionsButton = GenericItemCardOptionsButton;
GenericItemCard.Value = GenericItemCardValue;
GenericItemCard.ValuesGroup = GenericItemCardValuesGroup;
GenericItemCard.LineBreak = GenericItemCardLineBreak;
GenericItemCard.HeadingValue = GenericItemCardHeadingValue;

export default GenericItemCard;
