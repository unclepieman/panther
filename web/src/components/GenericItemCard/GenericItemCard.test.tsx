/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { render } from 'test-utils';
import logo from 'Assets/aws-minimal-logo.svg';
import { Box, Flex } from 'pouncejs';
import GenericItemCard from './index';

describe('GenericItemCard', () => {
  it('renders a step', () => {
    const { container } = render(
      <GenericItemCard>
        <GenericItemCard.Logo src={logo} />
        <GenericItemCard.Body>
          <GenericItemCard.ValuesGroup>
            <GenericItemCard.Value label="Test label" value="String value" />
            <GenericItemCard.Value label="Test label" value="Another value" />
            <GenericItemCard.LineBreak />
            <GenericItemCard.Value
              label="Test tailormade component"
              value={<Box>Cool right</Box>}
            />
            <Flex ml="auto" mr={0} align="flex-end">
              <Box>Test custom children</Box>
            </Flex>
            <GenericItemCard.Value value="without label" />
          </GenericItemCard.ValuesGroup>
        </GenericItemCard.Body>
      </GenericItemCard>
    );

    expect(container).toMatchSnapshot();
  });
});
