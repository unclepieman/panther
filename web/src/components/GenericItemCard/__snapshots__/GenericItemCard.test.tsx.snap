// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericItemCard renders a step 1`] = `
.panther-20 {
  background-color: #1d2a3f;
  border-radius: 4px;
  overflow: hidden;
}

.panther-19 {
  background-color: transparent;
  padding: 16px;
}

.panther-18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 100%;
}

.panther-0 {
  max-width: 100%;
  margin-right: 20px;
}

.panther-17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  width: 100%;
}

.panther-16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.panther-16>*:not(:last-child) {
  margin-right: 32px;
}

.panther-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-top: 8px;
}

.panther-1 {
  color: #bdbdbd;
  font-size: 0.563rem;
  font-weight: 500;
}

.panther-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font-size: 0.875rem;
  font-weight: 500;
  opacity: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 24px;
}

.panther-7 {
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  height: 0;
}



.panther-13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  margin-left: auto;
  margin-right: 0;
}

<div>
  <header
    id="main-header"
  />
  <section
    class="panther-20"
  >
    <div
      class="panther-19"
    >
      <div
        class="panther-18"
      >
        <img
          alt="Logo"
          class="panther-0"
          height="20"
          src="test-file-stub"
          width="20"
        />
        <div
          class="panther-17"
        >
          <div
            class="panther-16"
          >
            <dl
              class="panther-3"
            >
              <dt
                aria-labelledby="test-labelstring-value"
                class="panther-1"
              >
                Test label
              </dt>
              <dd
                aria-labelledby="test-labelstring-value"
                class="panther-2"
              >
                String value
              </dd>
            </dl>
            <dl
              class="panther-3"
            >
              <dt
                aria-labelledby="test-labelanother-value"
                class="panther-1"
              >
                Test label
              </dt>
              <dd
                aria-labelledby="test-labelanother-value"
                class="panther-2"
              >
                Another value
              </dd>
            </dl>
            <div
              class="panther-7"
            />
            <dl
              class="panther-3"
            >
              <dt
                aria-labelledby="test-tailormade-componentobject-object"
                class="panther-1"
              >
                Test tailormade component
              </dt>
              <dd
                aria-labelledby="test-tailormade-componentobject-object"
                class="panther-2"
              >
                <div
                  class="panther-9"
                >
                  Cool right
                </div>
              </dd>
            </dl>
            <div
              class="panther-13"
            >
              <div
                class="panther-9"
              >
                Test custom children
              </div>
            </div>
            <dl
              class="panther-3"
            >
              <dd
                aria-labelledby="undefinedwithout-label"
                class="panther-2"
              >
                without label
              </dd>
            </dl>
          </div>
        </div>
      </div>
    </div>
  </section>
  <footer
    id="footer"
  />
</div>
`;
